[SPARK-46302][TESTS] Skip maven daily testing as ivy uses some corrupted cache jar files

panbingkun opened the pull request:
### What changes were proposed in this pull request?
The pr aims to skip maven daily testing as ivy uses some corrupted cache jar files.
This is a temporary workaround solution.
After SPARK-46400 and applying it to the already released Spark version, we should remove this logic.

### Why are the changes needed?
Fix maven daily testing GA.
In our Maven daily testing, some UTs failed due to some corrupt jars in maven repo:
https://github.com/apache/spark/actions/runs/7019155617/job/19095991788#step:9:27263
<img width="993" alt="image" src="https://github.com/apache/spark/assets/15246973/def1e393-d0af-47b4-86d8-8c53a57d1ae0">
<img width="742" alt="image" src="https://github.com/apache/spark/assets/15246973/0955c447-20db-4157-8795-b279edfabd43">


### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA.
- Manually check.
<img width="986" alt="image" src="https://github.com/apache/spark/assets/15246973/c4ee3df3-f04e-438e-ac4d-6c6e50daf787">

### Was this patch authored or co-authored using generative AI tooling?
No.

panbingkun committed:
try fix maven test

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
fix maven tests

panbingkun committed:
Merge branch 'master' into try_fix_maven_test

panbingkun committed:
fix maven tests

panbingkun committed:
fix maven tests

panbingkun committed:
fix maven tests

panbingkun commented:
<img width="648" alt="image" src="https://github.com/apache/spark/assets/15246973/fb28e61c-510c-4e4e-b924-4f99cc7fe29e">

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
Merge branch 'master' into try_fix_maven_test

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
try

panbingkun committed:
Merge branch 'master' into try_fix_maven_test

panbingkun commented:
This is a workaround solution to make `HiveExternalCatalogVersionsSuite` test to pass.
- The root cause is that in this test, it relies on `hive-exec-2.3.9.jar` and `slf4j-api-1.7.30.jar`, but in our GA environment, these two files are incomplete, 
<img width="648" alt="image" src="https://github.com/apache/spark/assets/15246973/fb28e61c-510c-4e4e-b924-4f99cc7fe29e">
and ivy cannot detect this integrity.  
https://github.com/databricks/spark-redshift/issues/244#issuecomment-347082455

It directly loads the file from the local maven location, causing the failure of `HiveExternalCatalogVersionsSuite`.

- Because `HiveExternalCatalogVersionsSuite` test downloads the ·published spark file· for testing, modifying the code will not affect it. Here, we can only delete the corrupted file before executing the UT, so that ivy cannot recognize it, and then ivy will actively download the two completed jars from the remote side, thus bypassing it.

panbingkun commented:
Friendly ping @HyukjinKwon @zhengruifeng @LuciferYang 

I have discussed with @LuciferYang at local , and although this approach seems a bit strange at the moment, I have not yet found a better way to make `our maven daily testing` run happy.

panbingkun commented on .github/workflows/build_and_test.yml:
```diff
@@ -1111,3 +1111,9 @@ jobs:
           cd ui-test
           npm install --save-dev
           node --experimental-vm-modules node_modules/.bin/jest
+
```
Only for GA test.

LuciferYang commented:
Wait, I found that if I delete the local `.m2` and `.ivy2` folders, then I can reproduce this test failure locally.

LuciferYang commented:
In other words, if we run this test case in a completely clean environment, 

```
mv ~/.m2 ~/m2
mv ~/.ivy2 ~/ivy2
build/mvn clean install -DskipTests -Phive
build/mvn clean install -Phive -Dtest=none -DwildcardSuites=org.apache.spark.sql.hive.HiveExternalCatalogVersionsSuite -pl sql/hive                                                                      
```

then it will fail the test.


![image](https://github.com/apache/spark/assets/1475305/26250c40-815a-487d-a905-b45896405a84)

LuciferYang commented:
https://github.com/apache/spark/pull/44208#issuecomment-1849595508

also cc @pan3793 @yaooqinn Do you have any suggestions on this issue?

LuciferYang reviewed (changes requested):
After some thought, I realized this is not a simple test issue, it might be a long-standing bug, possibly existing since Spark started using the `hive-exec` with the `core` classifier. This could lead to job submission failures in specific scenarios when users submit jobs using the configuration `spark.sql.hive.metastore.jars=maven`.

Here are reproduce steps:

1. Execute `rm -rf ~/.m2` and `rm -rf ~/.ivy2` to get a clean test environment.
2. Build a Spark client using `dev/make-distribution.sh --tgz -Phive`, for example, I chose `3.4.2`.
3. Unzip the build result and execute the following test command:


```
bin/spark-submit '--name' 'prepare testing tables' '--master' 'local[2]' '--conf' 'spark.ui.enabled=false' '--conf' 'spark.master.rest.enabled=false' '--conf' 'spark.sql.hive.metastore.version=2.3' '--conf' 'spark.sql.hive.metastore.jars=maven' '--conf' 'spark.sql.warehouse.dir=/Users/yangjie01/Tools/spark-3.4.2-bin-3.3.4/test-warehouse' '--conf' 'spark.sql.test.version.index=0' '--driver-java-options' '-Dderby.system.home=/Users/yangjie01/Tools/spark-3.4.2-bin-3.3.4/test-warehouse' './test.py'
```

This test command and [test.py](https://github.com/apache/spark/files/13649551/test.txt) can be generated using `HiveExternalCatalogVersionsSuite`.

Afterwards, we can find that the submitted job has similar errors to the maven daily test:

```
:: problems summary ::
:::: WARNINGS
		[NOT FOUND  ] org.apache.hive#hive-exec;2.3.9!hive-exec.jar (3ms)

	==== local-m2-cache: tried

	  file:/Users/yangjie01/.m2/repository/org/apache/hive/hive-exec/2.3.9/hive-exec-2.3.9.jar

		::::::::::::::::::::::::::::::::::::::::::::::

		::              FAILED DOWNLOADS            ::

		:: ^ see resolution messages for details  ^ ::

		::::::::::::::::::::::::::::::::::::::::::::::

		:: org.apache.hive#hive-exec;2.3.9!hive-exec.jar

		::::::::::::::::::::::::::::::::::::::::::::::

```


The complete test log is as follows: [spark-submit.log](https://github.com/apache/spark/files/13649560/spark-submit.txt)

If `rm -rf ~/.m2` and `rm -rf ~/.ivy2` are executed again before executing step 3 to clear the environment, then the spark job can be executed successfully. I observed the generated `ivydata-2.3.9.properties`, if the corresponding `location` property points to the local `local-m2-cache` path, then the job submission will fail, if it points to the remote repository address, the test will succeed.

My suggestions are as follows:

1. For the current PR, an environment variable (such as a not test list) can be added to skip the problematic versions in the maven daily test (all currently released versions have this issue...)

2. For future fixes, there might be two possible approaches:

2.1 Solve through configuration: Add a configuration/system property/environment variable to forcibly skip the `local-m2-cache` `IBiblioResolver`.

2.2 Solve by Ivy API: For example, would supporting four-segment dependencies (`groupId:artifactId:classifier:version`) solve the problem? Or is there an API that can ignore `local-m2-cache` and retry remote dependencies in the above scenario? This might require some investigation work.

friendly ping @HyukjinKwon @dongjoon-hyun @srowen @JoshRosen  @yaooqinn @panbingkun Do you have any suggestions for this issue?  Thanks ~


To add, I have tested with Spark 3.1.3, and the aforementioned issue also exists.

panbingkun commented:
I don't think a four part approach may necessarily solve the above problem, because in our testing, we found another jar: `org.slf4j#slf4j-api#1.7.30")` with a similar problem that does not belong to this situation.

Let's continue to investigate it in depth.

panbingkun commented:
Based on the above discussion,

1. For the testing of already released files, we will skip it.
2. In response to the issue with `the master`, we choose to use a mechanism of `removing local Maven repo` as the cache after discovering that cache files have been corrupted, allowing Spark to automatically attempt it.

panbingkun committed:
[SPARK-46302][BUILD] Fix maven daily testing

panbingkun committed:
Merge branch 'master' into try_fix_maven_test

panbingkun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala:
```diff
@@ -212,45 +219,49 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
-      // Extract major.minor for testing Spark 3.1.x and 3.0.x with metastore 2.3.9 and Java 11.
-      val hiveMetastoreVersion = """^\d+\.\d+""".r.findFirstIn(hiveVersion).get
+  test("backward compatibility") {
+    if (!skipReleaseVersions.contains("master")) {
```
After `the master` is fixed, this condition can be removed.

panbingkun committed:
[SPARK-46302][BUILD] Fix maven daily testing

panbingkun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala:
```diff
@@ -66,6 +66,13 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
   val hiveVersion = HiveUtils.builtinHiveVersion
+  private val skipReleaseVersions =
+    sys.env.getOrElse("SKIP_SPARK_RELEASE_VERSIONS", "").split(",").toSet
+  // scalastyle:off println
```
Only for GA test.

panbingkun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala:
```diff
@@ -212,45 +219,54 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
-      val sparkHome = new File(sparkTestingDir, s"spark-$version")
-      if (!sparkHome.exists()) {
-        tryDownloadSpark(version, sparkTestingDir.getCanonicalPath)
+      // scalastyle:off println
```
Only for GA test.

panbingkun commented:
In this PR, some version tests have been skipped through environment variables.
After obtaining "Approve", I will remove some logic added for testing and observation.

LuciferYang commented:
How about just add a filter at 

https://github.com/apache/spark/blob/8cd466136528e479514a1d8dfe8ecaba9d0f8cce/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala#L256-L266

like:

```
val skipReleaseVersions =
    sys.env.getOrElse("SKIP_SPARK_RELEASE_VERSIONS", "3.3.4,3.4.2,3.5.0").split(",").toSet
...

source.mkString
        .split("\n")
        .filter(_.contains("""<a href="spark-"""))
        .filterNot(_.contains("preview"))
        .map("""<a href="spark-(\d.\d.\d)/">""".r.findFirstMatchIn(_).get.group(1))
        .filter(_ < org.apache.spark.SPARK_VERSION)
        .filterNot(skipReleaseVersions.contains).toImmutableArraySeq
```

panbingkun committed:
[SPARK-46302][BUILD] Fix maven daily testing

panbingkun committed:
Merge branch 'master' into try_fix_maven_test

panbingkun commented:
> How about just add a filter at
> 
> https://github.com/apache/spark/blob/8cd466136528e479514a1d8dfe8ecaba9d0f8cce/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala#L256-L266
> 
> like:
> 
> ```scala
> val skipReleaseVersions =
>     sys.env.getOrElse("SKIP_SPARK_RELEASE_VERSIONS", "").split(",").toSet
> ...
> 
> source.mkString
>         .split("\n")
>         .filter(_.contains("""<a href="spark-"""))
>         .filterNot(_.contains("preview"))
>         .map("""<a href="spark-(\d.\d.\d)/">""".r.findFirstMatchIn(_).get.group(1))
>         .filter(_ < org.apache.spark.SPARK_VERSION)
>         .filterNot(skipReleaseVersions.contains).toImmutableArraySeq
> ```

Done.

LuciferYang commented on .github/workflows/maven_test.yml:
```diff
@@ -40,7 +40,7 @@ on:
         required: false
         type: string
-        default: '{}'
+        default: '{"SKIP_SPARK_RELEASE_VERSIONS": "3.3.3,3.4.2,3.5.0,master"}'
```
No need for `master`

LuciferYang replied:
and default should be '{}', we should pass it from `build_maven.yml`

LuciferYang commented on .github/workflows/maven_test.yml:
```diff
@@ -115,6 +115,7 @@ jobs:
       HIVE_PROFILE: ${{ matrix.hive }}
       SPARK_LOCAL_IP: localhost
       GITHUB_PREV_SHA: ${{ github.event.before }}
+      SKIP_SPARK_RELEASE_VERSIONS: ${{ fromJSON(inputs.envs).SKIP_SPARK_RELEASE_VERSIONS }}
```
This option should not be needed, the `Run tests` section has already been deserialized for env

https://github.com/apache/spark/blob/0b8e0998c67331bc82428d807476584f3d21f934/.github/workflows/maven_test.yml#L172-L173

panbingkun committed:
[SPARK-46302][BUILD] Fix maven daily testing

LuciferYang commented on .github/workflows/build_maven.yml:
```diff
@@ -30,3 +30,8 @@ jobs:
+    with:
+      envs: >-
+        {
+          "SKIP_SPARK_RELEASE_VERSIONS": "3.3.3,3.4.2,3.5.0"
```
```suggestion
          "SKIP_SPARK_RELEASE_VERSIONS": "3.3.4,3.4.2,3.5.0"
```

panbingkun replied:
Done

panbingkun committed:
[SPARK-46302][BUILD] Fix maven daily testing

LuciferYang commented:
I think we should make https://github.com/apache/spark/pull/44343 as TODO of this one, and explicitly add TODO in the code. On the other hand, please update the PR title and PR description to better match the change of this PR.

panbingkun committed:
[SPARK-46302][TESTS] Skip maven daily testing as ivy uses some corrupted cache jar files

panbingkun commented:
> I think we should make #44343 as TODO of this one, and explicitly add TODO in the code. On the other hand, please update the PR title and PR description to better match the change of this PR.

Done.

LuciferYang commented on .github/workflows/build_and_test.yml:
```diff
@@ -1098,3 +1098,14 @@ jobs:
           npm install --save-dev
           node --experimental-vm-modules node_modules/.bin/jest
+
+  maven-test:
```
fine to me, let's revert this change after test pass

panbingkun replied:
Okay.

panbingkun replied:
I will revert this change, because the maven test has passed and the verification is successful.

panbingkun replied:
Done.

LuciferYang commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala:
```diff
@@ -244,11 +244,16 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
 }
 
 object PROCESS_TABLES extends QueryTest with SQLTestUtils {
+  // TODO After SPARK-46400 and applying it to the already released Spark version,
```
hmm... I think maybe we can describe it like
```
In SPARK-46302, the env SKIP_SPARK_RELEASE_VERSIONS has been added to allow Maven tests to skip problematic release versions. Related issues will be fixed in SPARK-46400, and testing will be resumed after the fixed Spark 3.x version is released.
```
and I think just add comments here, no need to add them in line 247

panbingkun replied:
Okay, let's run this test first.

in addition, before we release `3. x`, should we test `test("backward compatibility")` after the `SPARK-46400` feature is fixed on the branch master?

LuciferYang replied:
I don't think it has ever tested master...

LuciferYang replied:
Before the release of the fixed version, Maven's daily tests can skip this case first.

panbingkun committed:
[SPARK-46302][TESTS] Skip maven daily testing as ivy uses some corrupted cache jar files

panbingkun committed:
Merge branch 'master' into try_fix_maven_test

LuciferYang reviewed (approved):
+1, LGTM

panbingkun committed:
[SPARK-46302][TESTS] Skip maven daily testing as ivy uses some corrupted cache jar files

LuciferYang commented:
Merged into master. Thanks @panbingkun 

Let's monitor next maven daily test ~
